src/bin/cargo/commands/remove.rs

```diff
@@ -131,39 +135,91 @@ fn gc_workspace(workspace: &Workspace<'_>, dependencies: &[String]) -> CargoResu
-
-    Ok(())
-}
-
-/// Get whether or not a dependency is depended upon in a workspace.
-fn dep_in_workspace(dep: &str, members: &[LocalManifest]) -> bool {
-    members.iter().any(|manifest| {
-        manifest.get_sections().iter().any(|(_, table)| {
-            table
-                .as_table_like()
-                .unwrap()
-                .get(dep)
-                .and_then(|t| t.get("workspace"))
-                .and_then(|v| v.as_bool())
-                .unwrap_or(false)
+    let dependencies = members
+        .iter()
+        .flat_map(|manifest| {
+            manifest.get_sections().into_iter().flat_map(|(_, table)| {
+                table
+                    .as_table_like()
+                    .unwrap()
+                    .iter()
+                    .map(|(key, item)| Dependency::from_toml(&manifest.path, key, item))
+                    .collect::<Vec<_>>()
+            })
         })
-    })
-}
+        .collect::<CargoResult<Vec<_>>>()?;
 
-/// Remove a dependency from a workspace manifest.
-fn remove_workspace_dep(dep: &str, ws_manifest: &mut toml_edit::Document) {
-    if let Some(toml_edit::Item::Table(table)) = ws_manifest
+    // clean up workspace.dependencies
+    if let Some(toml_edit::Item::Table(deps_table)) = manifest
         .get_mut("workspace")
         .and_then(|t| t.get_mut("dependencies"))
     {
+        deps_table.set_implicit(true);
+        for (key, item) in deps_table.iter_mut() {
+            if !dependencies.iter().any(|d| {
+                d.toml_key() == key.get() && matches!(d.source(), Some(Source::Workspace(_)))
+            }) {
+                *item = toml_edit::Item::None;
+            }
+        }
+    }
+
+    // clean up the profile section
+    if let Some(toml_edit::Item::Table(profile_section_table)) = manifest.get_mut("profile") {
+        profile_section_table.set_implicit(true);
+        for (_, item) in profile_section_table.iter_mut() {
+            if let toml_edit::Item::Table(profile_table) = item {
+                profile_table.set_implicit(true);
+                if let Some(toml_edit::Item::Table(package_table)) =
+                    profile_table.get_mut("package")
+                {
+                    package_table.set_implicit(true);
+                    for (key, item) in package_table.iter_mut() {
+                        if let Ok(spec) = PackageIdSpec::parse(key.get()) {
+                            if !dependencies.iter().any(|d| d.name == spec.name().as_str()) {
+                                *item = toml_edit::Item::None;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // clean up patch section
+    if let Some(toml_edit::Item::Table(patch_section_table)) = manifest.get_mut("patch") {
+        patch_section_table.set_implicit(true);
+        for (_, item) in patch_section_table.iter_mut() {
```

Comment:
> The [patch] table is made of dependency-like sub-tables. Each key after [patch] is a URL of the source that is being patched, or the name of a registry.

Sounds like the key (`_`) is a source.  Should we be matching that with the source in the `Dependency`?

Reply:
Yes, we should. Fixed.